### PR TITLE
`fieldnames()` can throw for types such as `typeof(+)`

### LINKED ISSUE #32558: tab-completion: Error in the keymap
```
julia> typeof(+).<tab>
┌ Error: Error in the keymap                                                                                            
│   exception =                                                                                                                               
│    ArgumentError: type does not have a definite number of fields                                                                            
│    Stacktrace:                                                                                                                             
│     [1] fieldcount(::Any) at ./reflection.jl:682                                                                                          
│     [2] fieldnames(::DataType) at ./reflection.jl:172                                                                                      
│     [3] complete_symbol(::String, ::getfield(REPL.REPLCompletions, Symbol("##21#25")), ::Module) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/REPLCompletions
.jl:157                                                                                                                                            
│     [4] completions(::String, ::Int64, ::Module) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/REPLCompletions.jl:705         
│     [5] completions(::String, ::Int64) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/REPLCompletions.jl:575                    
│     [6] complete_line(::REPL.REPLCompletionProvider, ::Any) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:362               
│     [7] complete_line(::REPL.LineEdit.PromptState, ::Any) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/LineEdit.jl:324   
│     [8] complete_line(::REPL.LineEdit.MIState) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/LineEdit.jl:315             
│     [9] edit_tab(::REPL.LineEdit.MIState, ::Any, ::Any) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.3/REPL/src/LineEdit.jl:1987 (repeats 2 times)
```

### PR DESCRIPTION
**Pull request by staticfloat**
Wrap its invocation here in a `try` block.  Fixes #32558

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Maybe better to fix `fieldnames` instead/as well?

**Comment by JeffBezanson** [comment]
```
julia> fieldnames(typeof(+))
()
```
?

**Comment by StefanKarpinski** [comment]
Then what was the actual trigger here?

**Comment by staticfloat** [comment]
I don't know how to construct the exact path that leads us to this error; you can see it by typing `typeof(+).<tab>` into the REPL.  You can also trigger it via `fieldnames(Type{typeof(+)})`, although I'm not 100% certain that's the same fundamental error.

**Comment by JeffBezanson** [comment]
It is possible for `fieldnames` to error, but it will only do that on abstract types. It would be good to understand what's happening in the REPL. Is it supposed to reach here with an abstract type? If so, we could add a check for concreteness.

**Comment by ararslan** [comment]
This is interesting:
```
julia> t = typeof(+)
typeof(+)

julia> t.<TAB>

abstract         isbitstype        isinlinealloc     llvm::StructType  names             size              uid
hasfreetypevars  isconcretetype    layout            mutable           ninitialized      super             zeroinit
instance         isdispatchtuple   llvm::DIType      name              parameters        types
```

**Comment by ararslan** [comment]
> You can also trigger it via `fieldnames(Type{typeof(+)})`

The value `t` in https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/REPLCompletions.jl#L123 becomes `Type{typeof(+)}`, so the call on the line touched by this PR indeed becomes `fieldnames(Type{typeof(+)})`.

**Comment by JeffBezanson** [comment]
Ah, ok, so if `t` has the form `Type{T}`, we should convert it to `typeof(T)` instead. After that, we should also check `isconcretetype(t)`, and use empty fieldnames if false.

**Comment by staticfloat** [comment]
@JeffBezanson I've updated this PR (and rebased it on top of `master`) does this do what you described above?

File: stdlib/REPL/src/REPLCompletions.jl [hunk 1]
```diff
@@ -153,11 +153,18 @@ function complete_symbol(sym, ffunc, context_module=Main)::Vector{Completion}
     else
         # Looking for a member of a type
         if t isa DataType && t != Any
-            fields = fieldnames(t)
-            for field in fields
-                s = string(field)
-                if startswith(s, name)
-                    push!(suggestions, FieldCompletion(t, field))
+            # Check for cases like Type{typeof(+)}
+            if t isa Type && !isconcretetype(t) && length(t.parameters) >= 1
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPLCompletions.jl]:
```suggestion
            if t isa DataType && t.name === Base._TYPE_NAME
```

File: stdlib/REPL/src/REPLCompletions.jl [hunk 2]
```diff
@@ -153,11 +153,18 @@ function complete_symbol(sym, ffunc, context_module=Main)::Vector{Completion}
...
+                t = t.parameters[1]
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPLCompletions.jl]:
```suggestion
                t = typeof(t.parameters[1])
```

**Review by JeffBezanson — APPROVED** [review]

File: stdlib/REPL/test/replcompletions.jl [hunk 3]
```diff
@@ -1049,3 +1049,9 @@ let s = "prevind(\"θ\",1,"
     @test r == 1:7
     @test s[r] == "prevind"
 end
+
+# Issue #32840
+let s = "typeof(+)."
+    c, r = test_complete_context(s)
+    @test isempty(c)
```
Comment (JeffBezanson) [inline stdlib/REPL/test/replcompletions.jl]: This won't be empty --- it should contain `fieldnames(DataType)`.

File: stdlib/REPL/test/replcompletions.jl [hunk 4]
```diff
@@ -103,6 +103,7 @@ let s = ""
 end
 
 let s = "using REP"
+    @show completions(s, lastindex(s))[1]
```
Comment (JeffBezanson) [inline stdlib/REPL/test/replcompletions.jl]: Should this be removed?

**Pull request merged**